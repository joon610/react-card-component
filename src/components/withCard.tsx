/** @jsx jsx */
import { css,jsx } from '@emotion/react'
import React from 'react'
import {css_elevation, css_hover_option} from './styles';
interface BasicCardProps {
  children: JSX.Element
  className?: string
  elevation?: elevationType
  radius?: string;
  hoverOption?: hoverMoveType | undefined;
  onClick?: (event: React.MouseEvent<HTMLElement>) => void | undefined;
}

type elevationType =  0 | 1 | 2
type hoverMoveType = 'up' | 'left' | `right` | `down`


const css_root = css`
  width: 100%;
  height:100%;
  border-radius: 5px;
`

const css_child = css`
  height:100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
`


const BasicCard = (Component:React.FC<{ children?:JSX.Element}>) => (props:BasicCardProps) => {
  const elevationValue:elevationType =  props?.elevation || 0;
  const css_hover = props?.hoverOption && css_hover_option[props?.hoverOption];
  return (
  <div css={[css_root,css_elevation[elevationValue],css_hover]} className={props?.className} onClick={props.onClick}>
    <div css={css_child}>
      <Component { ...props}/>
    </div>
  </div>
  )
}

export default BasicCard;