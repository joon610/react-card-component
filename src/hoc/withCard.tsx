/** @jsx jsx */
import { css,jsx } from '@emotion/react'
import React from 'react'
import {css_elevation,css_root,css_child, css_hover_option} from './styles';

interface BasicCardProps {
  children: JSX.Element
  className?: string
  elevation?: elevationType
  radius?: string;
  hoverType?: hoverMoveType | undefined;
  onClick?: (event: React.MouseEvent<HTMLElement>) => void | undefined;
}

type elevationType =  0 | 1 | 2
type hoverMoveType = 'up' | 'left' | `right` | `down` | 'zoom'

const withCard = (Component:React.FC<{ children?:JSX.Element}>) => (props:BasicCardProps) => {
  const elevationValue:elevationType =  props?.elevation || 0;
  const css_hover = props?.hoverType && css_hover_option[props?.hoverType];
  const css_radius = props?.radius && css`border-radius:${props?.radius};`
  return (
  <div css={[css_root,css_elevation[elevationValue],css_hover,css_radius]} className={props?.className} onClick={props.onClick}>
    <div css={css_child}>
      <Component { ...props}/>
    </div>
  </div>
  )
}

export default withCard;