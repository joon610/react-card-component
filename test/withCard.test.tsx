import React from 'react'
import { fireEvent, render } from "@testing-library/react";
import { matchers } from '@emotion/jest'
import '@testing-library/jest-dom'
import withCard from '../src/components/withCard'
expect.extend(matchers)

describe('HOC writchCard',()=>{
  const ChildComponent = (props:any) => (<div>{props.children}</div>);
  const Component = withCard(ChildComponent);

  it('empty snapshot', () => {
    const {container} = render(<Component><div>good</div></Component>);
    expect(container).toMatchSnapshot();
  });

  it('Rendering Test', ()=>{
    const {getByText} = render(<Component><div>good</div></Component>);
    const childText = getByText('good');
    expect(childText).toBeInTheDocument();
  })

  it('click Card', ()=>{
    const handleClick = jest.fn();
    const { container } =render(<Component  onClick={handleClick}><div>good</div></Component>);
    const firstChild = container.getElementsByTagName('div')[0];
    fireEvent.click(firstChild);
    expect(firstChild).toHaveStyle('cursor:pointer')
    expect(handleClick).toHaveBeenCalledTimes(1);
  })

  it('radius',()=>{
    const { container } =render(<Component radius={"20px"}><div>good</div></Component>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyle({borderRadius:'20px'})
    expect(container).toMatchSnapshot();
  })

  it('elevationType 0',()=>{
    const { container } =render(<Component elevation={0}><div>good</div></Component>);
    expect(container).toMatchSnapshot();
  })

  it('elevationType 1',()=>{
    const { container } =render(<Component elevation={1}><div>good</div></Component>);
    expect(container).toMatchSnapshot();
  })

  it('elevationType 2',()=>{
    const { container } =render(<Component elevation={2}><div>good</div></Component>);
    expect(container).toMatchSnapshot();
  })

  it('hover zoom', async ()=>{
    const { container } =render(<Component hoverType={"zoom"}><div>good</div></Component>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyleRule("z-index", "10", { target: ":hover" });
    expect(firstChild).toHaveStyleRule("transform", "scale(1.2)", { target: ":hover" });
  })

  it('hover up', async ()=>{
    const { container } =render(<Component hoverType={"up"}><div>good</div></Component>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyleRule("transform", "translate(0px, -4px)", { target: ":hover" });
  })

  it('hover down', async ()=>{
    const { container } =render(<Component hoverType={"down"}><div>good</div></Component>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyleRule("transform", "translate(0px, 4px)", { target: ":hover" });
  })

  it('hover left', async ()=>{
    const { container } =render(<Component hoverType={"left"}><div>good</div></Component>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyleRule("transform", "translate(-4px, 0px)", { target: ":hover" });
  })

  it('hover right', async ()=>{
    const { container } =render(<Component hoverType={"right"}><div>good</div></Component>);
    const firstChild = container.getElementsByTagName('div')[0];
    expect(firstChild).toHaveStyleRule("transform", "translate(4px, 0px)", { target: ":hover" });
  })
})
