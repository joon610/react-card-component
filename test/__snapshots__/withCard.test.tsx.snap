// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOC writchCard elevationType 0 1`] = `
.emotion-0 {
  width: 100%;
  height: 100%;
  border-radius: 5px;
  background: white;
  box-shadow: none;
}

.emotion-1 {
  height: 100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div>
        <div>
          good
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`HOC writchCard elevationType 1 1`] = `
.emotion-0 {
  width: 100%;
  height: 100%;
  border-radius: 5px;
  background: white;
  box-shadow: 0px 2px 1px -1px rgb(0 0 0 / 20%),0px 1px 1px 0px rgb(0 0 0 / 14%),0px 1px 3px 0px rgb(0 0 0 / 12%);
}

.emotion-1 {
  height: 100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div>
        <div>
          good
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`HOC writchCard elevationType 2 1`] = `
.emotion-0 {
  width: 100%;
  height: 100%;
  border-radius: 5px;
  background: white;
  box-shadow: 0px 3px 3px -2px rgb(0 0 0 / 20%),0px 3px 4px 0px rgb(0 0 0 / 14%),0px 1px 8px 0px rgb(0 0 0 / 12%);
}

.emotion-1 {
  height: 100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div>
        <div>
          good
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`HOC writchCard empty snapshot 1`] = `
.emotion-0 {
  width: 100%;
  height: 100%;
  border-radius: 5px;
  background: white;
  box-shadow: none;
}

.emotion-1 {
  height: 100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div />
    </div>
  </div>
</div>
`;

exports[`HOC writchCard radius 1`] = `
.emotion-0 {
  width: 100%;
  height: 100%;
  border-radius: 5px;
  background: white;
  box-shadow: none;
  border-radius: 20px;
}

.emotion-1 {
  height: 100%;
  position: relative;
  border-radius: inherit;
  overflow: hidden;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div>
        <div>
          good
        </div>
      </div>
    </div>
  </div>
</div>
`;
